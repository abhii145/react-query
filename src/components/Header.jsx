import { Link, useNavigate, useLocation } from "react-router"
import { AiOutlineShoppingCart } from "react-icons/ai"
import { MdOutlineFavoriteBorder } from "react-icons/md"
import { useState, useCallback, useEffect } from "react"
import { debounce } from "lodash"
import { useSelector } from "react-redux"
import { FaBox } from "react-icons/fa6"

const Header = () => {
  const navigate = useNavigate()
  const location = useLocation()
  const [searchQuery, setSearchQuery] = useState("")
  const cartItems = useSelector((state) => state.cart.items)
  const cartItemCount = cartItems?.length

  const handleSearchChange = useCallback(
    debounce((value) => {
      navigate(`/products?search=${value}`)
    }, 500),
    [navigate]
  )

  const onSearchInputChange = (e) => {
    const value = e.target.value
    setSearchQuery(value)
    handleSearchChange(value)
  }

  useEffect(() => {
    const params = new URLSearchParams(location.search)
    const search = params.get("search")
    setSearchQuery(search || "")
  }, [location])

  return (
    <header className="bg-white shadow p-4 flex justify-between items-center">
      <Link to="/" className="text-xl font-bold" data-testid="logo-link">
        Logo
      </Link>
      <div className="flex space-x-4 items-center">
        <input
          type="text"
          value={searchQuery}
          onChange={onSearchInputChange}
          placeholder="Search products..."
          className="p-2 border border-gray-300 rounded-md text-sm focus:ring-2 focus:ring-blue-500 focus:outline-none"
        />
        <Link
          to="/favourite"
          className="text-xl font-bold"
          data-testid="favourite-link"
          aria-label="Favourite"
        >
          <MdOutlineFavoriteBorder className="text-2xl cursor-pointer fill-red-400" />
        </Link>

        <Link
          to="/order"
          className="text-xl font-bold"
          data-testid="order-link"
          aria-label="Order"
        >
          <FaBox className="text-2xl cursor-pointer fill-green-400" />
        </Link>

        <div className="relative">
          <Link to="/cart" data-testid="cart-link" aria-label="Cart">
            <AiOutlineShoppingCart className="text-2xl cursor-pointer fill-grey-100" />
            {cartItemCount > 0 && (
              <span className="bg-red-500 text-white text-xs absolute top-0 right-0 rounded-full w-4 h-4 flex items-center justify-center transform translate-x-1/2 -translate-y-1/2">
                {cartItemCount}
              </span>
            )}
          </Link>
        </div>
      </div>
    </header>
  )
}

export default Header
