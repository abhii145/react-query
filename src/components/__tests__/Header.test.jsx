import { render, screen, fireEvent } from "@testing-library/react"
import { MemoryRouter } from "react-router-dom"
import Header from "../Header"
import { describe, it, expect, vi, beforeEach } from "vitest"
import { Provider } from "react-redux"
import { configureStore } from "@reduxjs/toolkit"
import cartReducer from "../../store/cartSlice"

const mockNavigate = vi.fn()

vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal()
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  }
})

describe("Header component", () => {
  let store

  beforeEach(() => {
    store = configureStore({
      reducer: {
        cart: cartReducer,
      },
      preloadedState: {
        cart: {
          items: [],
        },
      },
    })
  })

  it("matches snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it("renders Header component", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </Provider>
    )

    expect(screen.getByText("Logo")).toBeInTheDocument()
    expect(screen.getByTestId("logo-link")).toBeInTheDocument()
    expect(screen.getByTestId("favourite-link")).toBeInTheDocument()
    expect(screen.getByTestId("cart-link")).toBeInTheDocument()
    expect(screen.getByTestId("order-link")).toBeInTheDocument()
  })

  it("links navigate to correct paths", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </Provider>
    )

    expect(screen.getByTestId("logo-link")).toHaveAttribute("href", "/")
    expect(screen.getByTestId("favourite-link")).toHaveAttribute(
      "href",
      "/favourite"
    )
    expect(screen.getByTestId("cart-link")).toHaveAttribute("href", "/cart")
    expect(screen.getByTestId("order-link")).toHaveAttribute("href", "/orders")
  })

  it("displays the correct number of cart items", () => {
    store = configureStore({
      reducer: {
        cart: cartReducer,
      },
      preloadedState: {
        cart: {
          items: [{ id: 1 }, { id: 2 }],
        },
      },
    })

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </Provider>
    )

    expect(screen.getByText("2")).toBeInTheDocument()
  })

  it("updates search query on input change", () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Header />
        </MemoryRouter>
      </Provider>
    )

    const searchInput = screen.getByPlaceholderText("Search products...")
    fireEvent.change(searchInput, { target: { value: "test" } })

    expect(searchInput.value).toBe("test")
  })

  it("displays the correct search query from URL", () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/products?search=example"]}>
          <Header />
        </MemoryRouter>
      </Provider>
    )

    const searchInput = screen.getByPlaceholderText("Search products...")
    expect(searchInput.value).toBe("example")
  })
})
