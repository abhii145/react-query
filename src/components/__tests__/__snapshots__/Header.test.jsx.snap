// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="bg-white shadow p-4 flex justify-between items-center"
  >
    <a
      class="text-xl font-bold"
      data-discover="true"
      data-testid="logo-link"
      href="/"
    >
      Logo
    </a>
    <div
      class="flex space-x-4 items-center"
    >
      <input
        class="p-2 border border-gray-300 rounded-md text-sm focus:ring-2 focus:ring-blue-500 focus:outline-none"
        placeholder="Search products..."
        type="text"
        value=""
      />
      <a
        aria-label="Favourite"
        class="text-xl font-bold"
        data-discover="true"
        data-testid="favourite-link"
        href="/favourite"
      >
        <svg
          class="text-2xl cursor-pointer fill-red-400"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
          />
        </svg>
      </a>
      <a
        aria-label="Cart"
        data-discover="true"
        data-testid="cart-link"
        href="/cart"
      >
        <svg
          class="text-2xl cursor-pointer fill-grey-100"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 1024 1024"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 0 0-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 1 0 0 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 0 0-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 0 0-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 0 0-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 0 0-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 0 1-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 0 1-31.6 31.6z"
          />
        </svg>
      </a>
    </div>
  </header>
</DocumentFragment>
`;
